core/state/pruner/pruner.go

```diff
@@ -228,10 +233,130 @@ func prune(snaptree *snapshot.Tree, root common.Hash, maindb ethdb.Database, sta
 	return nil
 }
 
+func nodeIteratorKey(it trie.NodeIterator) (common.Hash, error) {
+	if it.Leaf() {
+		keyBytes := it.LeafKey()
+		if len(keyBytes) != len(common.Hash{}) {
+			return common.Hash{}, fmt.Errorf("unexpected db key length %v", len(keyBytes))
+		}
+		key := common.BytesToHash(keyBytes)
+		if key == (common.Hash{}) {
+			return common.Hash{}, errors.New("got zero key for trie db node")
+		}
+		return key, nil
+	} else {
+		return it.Hash(), nil
+	}
+}
+
+// We assume output does not need the value, only the key
+func dumpRawTrieDescendants(db ethdb.Database, root common.Hash, output ethdb.KeyValueWriter) error {
```

Comment:
taking a stateBloo instead of KeyValueWriter would be better for self-documentation